### DISCUSSION TITLE
loading certificates and keys in a privilege-separated unprivileged process

### DISCUSSION THREAD

**Discussion by matthewluckie** [openssl/openssl#24956] [Q&A] [answered]
Hi,

I would like to periodically re-read certificates and keys in a privilege separated program.  The unprivileged process, which is chroot'd and so does not have access to the files containing the material itself, would need assistance from the privileged process that does have access to the material.  These are keys that automatically roll every 90 days.  I usually call 

SSL_CTX_use_certificate_chain_file
SSL_CTX_use_PrivateKey_file

My mental model for how I might solve this is to read the contents of these files in the privileged process, and sendmsg it to the unprivileged process.  Is there a set of routines that would help take that raw buffer in the unprivileged process and load it into the SSL_CTX?  Do I essentially replicate the code in

https://github.com/openssl/openssl/blob/a4fd94851261c55f9ad020bf22d4f29bda0b58be/ssl/ssl_rsa.c#L437

replacing BIO_s_file() with a BIO_s_mem() ?

Matthew

**Comment by t8m** [comment, marked as answer]
You can use SSL_CTX_use_cert_and_key() and parse the key and certificate with the respective PEM_read_BIO functions.
